sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java

```diff
@@ -0,0 +1,357 @@
+                expression.startCollect(collectorContext);
+            }
+            InputRow inputRow = new InputRow(docCtx.topLevelInputs());
+
+            LuceneQueryBuilder.Context queryContext = luceneQueryBuilder.convert(
+                collectPhase.where(),
+                indexShard.mapperService(),
+                queryShardContext,
+                sharedShardContext.indexService().cache()
+            );
+            return CollectingBatchIterator.newInstance(
+                searcher::close,
+                t -> {},
+                () -> {
+                    try {
+                        return CompletableFuture.completedFuture(
+                            getRows(
+                                getGroupedStates(
+                                    bigArrays,
+                                    searcher,
+                                    keyIndexFieldData,
+                                    ctxForAggregations,
+                                    aggregations,
+                                    expressions,
+                                    ramAccounting,
+                                    inputRow,
+                                    queryContext
+                                ),
+                                ramAccounting,
+                                aggregations,
+                                groupProjection.mode()
+                            )
+                        );
+                    } catch (Throwable t) {
+                        return failedFuture(t);
+                    }
+                }
+            );
+        } catch (Throwable t) {
+            searcher.close();
+            throw t;
+        }
+    }
+
+    private static Iterable<Row> getRows(Map<BytesRef, Object[]> groupedStates,
+                                         RamAccountingContext ramAccounting,
+                                         List<AggregationContext> aggregations,
+                                         AggregateMode mode) {
+        return () -> groupedStates.entrySet().stream()
+            .map(new Function<Map.Entry<BytesRef, Object[]>, Row>() {
+
+                final Object[] cells = new Object[1 + aggregations.size()];
+                final RowN row = new RowN(cells);
+
+                @Override
+                public Row apply(Map.Entry<BytesRef, Object[]> entry) {
+                    cells[0] = entry.getKey();
+                    Object[] states = entry.getValue();
+                    for (int i = 0, c = 1; i < states.length; i++, c++) {
+                        //noinspection unchecked
+                        cells[c] = mode.finishCollect(ramAccounting, aggregations.get(i).function(), states[i]);
+                    }
+                    return row;
+                }
+            })
+            .iterator();
+    }
+
+    private static Map<BytesRef, Object[]> getGroupedStates(BigArrays bigArrays,
```

Comment:
can we document this? (or rename it to something that indicates what kind of grouping it's performing - groupedStatesByOrdinalRef maybe? - or something along those lines)